### PR TITLE
gomplate 2.0.0 (new formula)

### PR DESCRIPTION
**Pull request by docwhat**
A small and powerful template renderer; extremely useful for system administration.

Documentation: https://gomplate.hairyhenderson.ca/
GitHub page: https://github.com/hairyhenderson/gomplate

### CODE REVIEW

**Comment by docwhat** [comment]
I get this warning from `brew audit` but I want to use the version number in the `url`.  Is there a better way to do this?

```
$ brew audit --new-formula gomplate
gomplate:
  * C: 5: col 3: `url` (line 5) should be put before `version` (line 4)
Error: 1 problem in 1 formula
```

**Comment by DomT4** [comment]
> I get this warning from brew audit but I want to use the version number in the url

Homebrew intentionally doesn't support this style. The version should be in the URL directly.

This formula will also need building from source, like the rest of the `go`-using formulae Homebrew provide. Since upstream vendor their dependencies that's nice & simple at least. Something like the [peco](https://github.com/Homebrew/homebrew-core/blob/46949a473d605ff1f243c0b8e5d7e04f218c19dc/Formula/peco.rb) formula probably fits the bill as a template.

**Comment by docwhat** [comment]
> Something like the peco formula probably fits the bill as a template.

Thanks!  The one (wonderful) problem I had was there are too many examples; I couldn't find the right one. 😃

**Comment by docwhat** [comment]
Okay, I rewrote it based on `peco` as an example.  Thanks for the feedback.

**Comment by docwhat** [comment]
> Homebrew intentionally doesn't support this style. The version should be in the URL directly.

I need the version number in multiple places.  I used a constant `VERSION`.  Is that acceptable? Or should I just hard-code it into the various places and require every change to fix all the references?

**Comment by docwhat** [comment]
>  I used a constant VERSION.

Never mind, I figured out I can use `version` even though I didn't declare it (it's auto-detected via the `url` I guess?).

**Comment by DomT4** [comment]
Yeah, `version` is automatically detected _(or at least, an attempt is made to detect it)_ and globally available, even outside of the individual formula. If you needed to, for example, something like `gcc_version = Formula["gcc"].version.to_s` would work perfectly well in any formula.

File: Formula/gomplate.rb
Comment (DomT4) [inline Formula/gomplate.rb]: You can squash these two lines IIRC into:

```ruby
bin.install "bin/gomplate-slim" => "gomplate"
```

Comment (DomT4) [inline Formula/gomplate.rb]: It's probably obvious, but "these two lines" == 18 & 19.

Comment (docwhat) [inline Formula/gomplate.rb]: Thanks! I knew that had to be there someplace, but the [documentation for Pathname#install](http://www.rubydoc.info/github/Homebrew/brew/master/Pathname#install-instance_method) doesn't mention that.

Comment (docwhat) [inline Formula/gomplate.rb]: Done.

File: Formula/gomplate.rb
Comment (DomT4) [inline Formula/gomplate.rb]: Why `pipe_output` here, out of curiosity? Does `--version` throw a "dirty" exit code?

**Comment by docwhat** [comment]
I wanted the output. I assumed it checked for exit.code 0.

Does it not?

**Comment by DomT4** [comment]
No, not automatically. `shell_output` checks for a `0` exit code by default, whereas `pipe_output` doesn't. If you need to use `pipe_output` you can make it check [like this](https://github.com/Homebrew/homebrew-core/blob/ba0aa68066a89c0654152aba22819736b06987af/Formula/repl.rb#L17) but otherwise changing to `shell_output` will do so for you.

**Comment by docwhat** [comment]
@DomT4: Thanks!  I didn't know that.  I have fixed my code.

File: Formula/gomplate.rb
Comment (DomT4) [inline Formula/gomplate.rb]: Need a `prefix.install_metafiles` after this line, but I think that's the last thing I can see that's problematic. After that it's down to one of the remaining maintainers to decide whether I've missed anything 😄.

Comment (docwhat) [inline Formula/gomplate.rb]: Done!

What does that do? Why didn't `brew audit` flag it?

Comment (DomT4) [inline Formula/gomplate.rb]: Heads up that the rebase went wrong here, you've ended up merging in all the other commits to master lately, heh.

`install_metafiles` means install `LICENSE`, `README`, and other related files. It normally happens automatically but for builds that `cd` somewhere else for the installation process it can be missed. There have been various discussions around improving that detection but no happy middle ground has been found, to date.

Comment (docwhat) [inline Formula/gomplate.rb]: Thanks. I fixed the rebase. D'oh!

**Comment by DomT4** [comment]
@ilovezfs I think this is in a reasonably good place if you're happy. There's some potentially redundant brackets inside the `test do` block around the first `assert_match` but 🤷‍♂️.

**Comment by docwhat** [comment]
> There's some potentially redundant brackets inside the test do block around the first assert_match but 🤷‍♂️ .

I've been writing in too many languages in the last week.  I recovered my Ruby mojo and removed the extra parens. 😉

File: Formula/gomplate.rb
Comment (DomT4) [inline Formula/gomplate.rb]: This really is anal on my part, but brackets are normal here. It was the line below I was gently referencing the other day.

```ruby
ver = shell_output("#{bin}/gomplate --version")
assert_match "gomplate version #{version}", ver
```
essentially.

**Review by DomT4 — APPROVED** [review]
One very minor nit that I can see but otherwise I think purely in terms of syntax this seems fine. Will leave this approval tick here for ILZ to find & agree/disagree with me about 😉.

**Comment by docwhat** [comment]
Do I need to do anything I can do to help get this accepted?

**Comment by DomT4** [comment]
I pointed out one style nit in a review comment above, but otherwise no, this needs reviewing by someone who can actually merge things these days. I can do review but proving perhaps I'd be a bad politician I gave up the actual power I had 😄.

Teeny tiny friendly nudge @fxcoudert @ilovezfs ❤️.

File: Formula/gomplate.rb
Comment (ilovezfs) [inline Formula/gomplate.rb]: what is the exit code?

File: Formula/gomplate.rb
Comment (ilovezfs) [inline Formula/gomplate.rb]: would prefer a heredoc above

File: Formula/gomplate.rb
Comment (ilovezfs) [inline Formula/gomplate.rb]: version test is unneeded if we're doing the functional test below

File: Formula/gomplate.rb
Comment (ilovezfs) [inline Formula/gomplate.rb]: can we do without the upx dep

File: Formula/gomplate.rb
Comment (ilovezfs) [inline Formula/gomplate.rb]: also prefer a heredoc

**Comment by hairyhenderson** [comment]
Thanks for submitting this @docwhat! Would be great to be in core. I already have a tap at https://github.com/hairyhenderson/homebrew-tap - though not sure that would be acceptable here.

File: Formula/gomplate.rb
Comment (hairyhenderson) [inline Formula/gomplate.rb]: `upx` is used by the `make compress` target

**Comment by hairyhenderson** [comment]
Just FYI, `gomplate` 1.10.0 is out now (https://github.com/hairyhenderson/gomplate/releases/tag/v1.10.0), and v2.0.0 will likely be released sometime in the next few days.

File: Formula/gomplate.rb
Comment (docwhat) [inline Formula/gomplate.rb]: Actually, that is to test that we compiled so that the version number is set. I'll add a comment to that effect.

File: Formula/gomplate.rb
Comment (docwhat) [inline Formula/gomplate.rb]: I'll add it; it expect 0.  I thought it would be the default.

File: Formula/gomplate.rb
Comment (docwhat) [inline Formula/gomplate.rb]: Done.

File: Formula/gomplate.rb
Comment (docwhat) [inline Formula/gomplate.rb]: It can be removed but it will increase the size of the cask.

UPX output:

```
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2017
UPX 3.94        Markus Oberhumer, Laszlo Molnar & John Reiser   May 12th 2017

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
  16043316 ->   4223632   26.33%   macho/amd64   gomplate-slim

Packed 1 file.
```

So 16mb is compressed to 4mb.

Your call: I can easily change it.

File: Formula/gomplate.rb
Comment (docwhat) [inline Formula/gomplate.rb]: done.

File: Formula/gomplate.rb
Comment (docwhat) [inline Formula/gomplate.rb]: done.

File: Formula/gomplate.rb
Comment (docwhat) [inline Formula/gomplate.rb]: added comment

**Comment by dunn** [comment]
@BrewTestBot test this please

**Comment by dunn** [comment]
Just a small audit failure:
```
Error: 1 problem in 1 formula
gomplate:
  * C: 28: col 1: Extra blank line detected.

File: Formula/gomplate.rb
Comment (DomT4) [inline Formula/gomplate.rb]: World's smallest nit for the sake of consistency:
```ruby
ver = shell_output("#{bin}/gomplate --version")
```

File: Formula/gomplate.rb
Comment (DomT4) [inline Formula/gomplate.rb]: @ilovezfs You missed the commas with your push 😉. This line should be:
```ruby
output = shell_output("#{bin}/gomplate --version")
```

**Comment by hairyhenderson** [comment]
Just FYI, the current gomplate release is now 2.0.0 (https://github.com/hairyhenderson/gomplate/releases/tag/v2.0.0). Might be useful to update this PR to reflect that.

**Review by ilovezfs — APPROVED** [review]

**Pull request closed without merging**
